### PR TITLE
Multi dense vector simple storage

### PR DESCRIPTION
**Pull request by agourlay**
A simple storage using RocksDB for multi vectors points.

### CODE REVIEW

File: lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            bincode::serialize(&key).unwrap(),
+            bincode::serialize(&record).unwrap(),
+        )?;
+
+        Ok(())
+    }
+}
+
+// TODO integrate MultiDenseVector to Vectors enum to enable this implementation
+// impl VectorStorage for SimpleMultiDenseVectorStorage
```
Comment (agourlay) [inline lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs:124]: @IvanPleshkov implementing this trait means introducing a new type in the Vector hierarchy.

We have seen that movie already, better do it separately :sweat_smile:

File: lib/segment/src/data_types/named_vectors.rs [hunk 2]
```diff
@@ -155,6 +175,7 @@ impl<'a> NamedVectors<'a> {
             match vector {
                 Vector::Dense(v) => CowVector::Dense(Cow::Owned(v)),
                 Vector::Sparse(v) => CowVector::Sparse(Cow::Owned(v)),
+                Vector::MultiDense(_) => unimplemented!("MultiDenseVector is not supported"),
```
Comment (IvanPleshkov) [inline lib/segment/src/data_types/named_vectors.rs:178]: Is it something missing here to implement?

Comment (agourlay) [reply lib/segment/src/data_types/named_vectors.rs:178]: Good catch, I will add the impl :+1:

Comment (hiepxanh) [reply lib/segment/src/data_types/named_vectors.rs:178]: @agourlay  can you take a look on my question? I'm trying to work with spare vector and wonder how to regenerate vector without insert the same paayload again 
https://github.com/orgs/qdrant/discussions/3797

Comment (generall) [reply lib/segment/src/data_types/named_vectors.rs:178]: Hey @hiepxanh, this place is not related to your question. Please keep the discussion within the expected topic.

Comment (hiepxanh) [reply lib/segment/src/data_types/named_vectors.rs:178]: Yes sir, I'm sorry

File: lib/segment/src/segment.rs [hunk 3]
```diff
@@ -211,11 +211,17 @@ impl Segment {
...
                         VectorStorageEnum::DenseSimple(_)
                         | VectorStorageEnum::DenseMemmap(_)
-                        | VectorStorageEnum::DenseAppendableMemmap(_) => vec![1.0; dim].into(),
-                        VectorStorageEnum::SparseSimple(_) => SparseVector::default().into(),
+                        | VectorStorageEnum::DenseAppendableMemmap(_) => {
+                            Vector::from(vec![1.0; dim])
+                        }
+                        VectorStorageEnum::SparseSimple(_) => Vector::from(SparseVector::default()),
+                        VectorStorageEnum::MultiDenseSimple(_) => {
+                            Vector::from(vec![vec![1.0; dim]])
```
Comment (IvanPleshkov) [inline lib/segment/src/segment.rs:223]: I don't know if empty multivectror is the correct case. If yes, I propose to use an empty vector

File: lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    dim: usize,
+    distance: Distance,
+    stopped: &AtomicBool,
+) -> OperationResult<Arc<AtomicRefCell<VectorStorageEnum>>> {
+    let mut vectors = vec![];
+    let (mut deleted, mut deleted_count) = (BitVec::new(), 0);
+    let db_wrapper = DatabaseColumnWrapper::new(database, database_column_name);
+
+    let mut total_vector_count = 0;
+    let mut total_sparse_size = 0;
```
Comment (IvanPleshkov) [inline lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs:51]: `total_sparse_size` is not required here

File: lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+            .map_err(|_| OperationError::service_error("cannot deserialize point id from db"))?;
+        let stored_record: StoredMultiDenseVector = bincode::deserialize(&value)
+            .map_err(|_| OperationError::service_error("cannot deserialize record from db"))?;
+
+        // Propagate deleted flag
+        if stored_record.deleted {
+            bitvec_set_deleted(&mut deleted, point_id, true);
+            deleted_count += 1;
+        }
+        vectors.insert(point_id as usize, stored_record.vector);
```
Comment (IvanPleshkov) [inline lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs:64]: `insert` panics when point_id >= length
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.insert
also it shifts elements

File: lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs (hunk 4 above)
Comment (IvanPleshkov) [reply lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs:51]: `total_sparse_size` is not requires too

File: lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+            deleted,
+            deleted_count,
+        }),
+    )))
+}
+
+impl SimpleMultiDenseVectorStorage {
+    /// Set deleted flag for given key. Returns previous deleted state.
+    #[inline]
+    #[allow(unused)]
```
Comment (IvanPleshkov) [inline lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs:88]: `#[allow(unused)]` can be removed

File: lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+            if !was_deleted {
+                self.deleted_count += 1;
+            } else {
+                self.deleted_count = self.deleted_count.saturating_sub(1);
+            }
+        }
+        was_deleted
+    }
+
+    #[allow(unused)]
```
Comment (IvanPleshkov) [inline lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs:104]: `#[allow(unused)]` can be removed

File: lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    fn total_vector_count(&self) -> usize {
+        self.vectors.len()
+    }
+
+    fn get_vector(&self, key: PointOffsetType) -> CowVector {
+        self.vectors
+            .get(key as usize)
+            .expect("vector not found")
+            .as_slice()
+            .into()
```
Comment (IvanPleshkov) [inline lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs:150]: I propose to use `from` =)

**Review by IvanPleshkov — CHANGES_REQUESTED** [review]

File: lib/segment/src/segment.rs (hunk 3 above)
Comment (IvanPleshkov) [reply lib/segment/src/segment.rs:223]: If seems multivector cannot be empty because score is not defined

Comment (agourlay) [reply lib/segment/src/segment.rs:223]: What do you think about my choice?

Comment (IvanPleshkov) [reply lib/segment/src/segment.rs:223]: There are no definition of empty multi vector score, empty multivector is invalid. It means your choice is excellent

**Review by IvanPleshkov — APPROVED** [review]

**Pull request merged**